### PR TITLE
Fixing cast of `n * (log(n) + log(log(n)))` in function `prime`

### LINKED ISSUE #25698: n=6000002; `int(n*(log(n) + log(log(n))))` takes more than 200 s to compute on sympy.live.org
Version: sympy 1.12

The `sympy.prime` takes an unnaturally long time to return its value on certain arguments, for example, $6000002$. Note that feeding surrounding values such as $6000001$ and $6000003$ to the function does not yield such a hang.

```
sympy.prime(6000001) # Instant
sympy.prime(6000002) # Takes well over a minute
sympy.prime(6000003) # Instant
```

This is likely due to the code of calculating the upper bound for binary searching the value of $\mathrm{li}^{-1}$.
https://github.com/sympy/sympy/blob/f59cbf85581d2b9cd23d8f864607f4520df2366f/sympy/ntheory/generate.py#L441
Here, the equation `n * (log(n) + log(log(n)))` is calculated, and then casted to an `int`. When $n$ is $6000002$, the value of `n * (log(n) + log(log(n)))` is close to an integer (`110130079.99980894` when casted to a float). 
When provided with an $n$ such that the value of $n (\log n + \log \log n)$ is close to an integer, the `sympy.prime` function becomes extremely slow compared to other values close to $n$.

### PR DESCRIPTION
**Pull request by platypus999**
#### References to other Issues or PRs

Fixes #25698 

#### Brief description of what is fixed or changed

Evaluates `n * (log(n) + log(log(n)))` to a numerical value then cast to integer, reducing lag in certain values of `n`.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Comment by smichr** [comment]
If this is accurate then it leaves the question: why doesn't `int` already do this? My guess is that one must calculate enough digits to know whether the number is going to round up or down. So if the decimal start '9999' you have to keep computing until you get a non-9 or else run out of resources. This number is not that big and I'm not sure why it is having issues. The fix here might be off by `n` assuming that the value in parenthesis might be off by `1`. So I am reluctant to consider this the correct fix.

**Comment by platypus999** [comment]
> The fix here might be off by n assuming that the value in parenthesis might be off by 1. So I am reluctant to consider this the correct fix.

If the equation is casted like `n * int(log(n) + log(log(n)))` then this might be an issue, but I believe that casting after the multiplication of `n` would just result in a value off by $O(1)$. 
And it seems like the parenthesis is expanded anyway when the sympy equation object is constructed.

```
n = 6000002
print(n * (sympy.log(n) + sympy.log(sympy.log(n))))
```

Output: `6000002*log(log(6000002)) + 6000002*log(6000002)`

**Comment by platypus999** [comment]
If we want to be extra safe, doing something like `b = int(2 * n * log(n))` will ensure that the value will be no less than the original value, and since `b` is just used as a boundary of a binary search algorithm this will only affect the running time by one additional run of the loop (because the new `b` is only twice as large as the original value).

**Comment by smichr** [comment]
I left comments on the issue page. The proper fix is to make `__int__` work properly.

File: sympy/ntheory/generate.py [hunk 1]
```diff
@@ -438,7 +438,7 @@ def prime(nth):
     from sympy.functions.elementary.exponential import log
     from sympy.functions.special.error_functions import li
     a = 2 # Lower bound for binary search
-    b = int(n*(log(n) + log(log(n)))) # Upper bound for the search.
+    b = int((n*(log(n) + log(log(n)))).n()) # Upper bound for the search.
```
Comment (smichr) [inline sympy/ntheory/generate.py:441]:
```suggestion
    b = n*int(log(n) + log(log(n))) # Upper bound for the search.
```

File: sympy/core/expr.py [hunk 2]
```diff
@@ -7,6 +7,7 @@
 
 from .sympify import sympify, _sympify
 from .basic import Basic, Atom
+from .numbers import int_valued
```
Comment (smichr) [inline sympy/core/expr.py:10]:
```suggestion
```

File: sympy/core/expr.py [hunk 3]
```diff
@@ -324,19 +311,13 @@ def __int__(self):
...
-            # was added when this comment was added passes, it might be safe
-            # to simply use sign to compute this rather than doing this by hand:
-            diff_sign = 1 if (self - x).evalf(2, subs={x: i}) > 0 else -1
-            if diff_sign != isign:
-                i -= isign
+        if int_valued(r):
+            # non-integer self should pass one of these tests
+            if self > i is S.true:
+                return i
+            if self < i is S.true:
```
Comment (smichr) [inline sympy/core/expr.py:318]:
```suggestion
            if (self > i) is S.true:
                return i
            if (self < i) is S.true:
```

**Comment by smichr** [comment]
More is needed from `__int__`. Consider

```
>>> eq = exp(I*pi*(cos(2)**2+sin(2)**2))
>>. eq.round(2)
-1.0 - 0.e-7*I
```
So perhaps
```
r, im = eq.round(2)
if im._prec == 1:
    if eq.equals(int(r)):
        return int(r)
```

**Comment by smichr** [comment]
@platypus999 , this was a great error report. I think we have a significant improvement. For sure in this specific case but also for other expressions like the one added in the tests. I will commit tomorrow if there is no further input and tests pass.

File: sympy/core/expr.py [hunk 4]
```diff
@@ -3920,6 +3904,11 @@ def round(self, n=None):
         # the 2nd digit from the left) we get 5700000000000000.
         #
         xf = x.n(dps + extra)*Pow(10, shift)
+        if xf._prec == 1:
```
Comment (smichr) [inline sympy/core/expr.py:3907]:
```suggestion
        if xf.is_Number and xf._prec == 1:  # case where xf.is_Add will raise below
```

File: sympy/ntheory/generate.py [hunk 5]
```diff
@@ -438,7 +438,7 @@ def prime(nth):
     from sympy.functions.elementary.exponential import log
     from sympy.functions.special.error_functions import li
     a = 2 # Lower bound for binary search
-    b = int(n*(log(n) + log(log(n)))) # Upper bound for the search.
```
Comment (haru-44) [inline sympy/ntheory/generate.py:441]: I don't think we need to make any fixes here.

Comment (smichr) [reply sympy/ntheory/generate.py:441]: Thanks - I made that change locally but didn't copy to here.

File: sympy/ntheory/generate.py [hunk 6]
```diff
@@ -438,7 +438,7 @@ def prime(nth):
...
+    b = n*int(log(n) + log(log(n))) # Upper bound for the search.
```
Comment (smichr) [inline sympy/ntheory/generate.py:441]:
```suggestion
    b = int(n*(log(n) + log(log(n)))) # Upper bound for the search.
```

File: sympy/core/tests/test_expr.py [hunk 7]
```diff
@@ -2084,6 +2084,15 @@ def test_round():
...
         assert S(fi).round().is_Integer
 
+        # issue 25698
+        one = cos(2)**2 + sin(2)**2
+        eq = exp(one*I*pi)
+        qr, qi = eq.as_real_imag()
+        assert qi.round(2) == 0.0
+        assert eq.round(2) == -1.0
+        n = 6000002
+        assert int(n*(log(n) + log(log(n)))) == 110130079
```
Comment (smichr) [inline sympy/core/tests/test_expr.py:2094]:
```suggestion
        assert int(n*(log(n) + log(log(n)))) == 110130079
        eq = one - 1/S(10**120)
        assert int(eq) == 0
        assert int(-eq) == 0
```

**Comment by asmeurer** [comment]
Shouldn't the code in round be handling this? The checks here imply that it's round that isn't correct.

File: sympy/core/expr.py [hunk 8]
```diff
@@ -324,19 +311,24 @@ def __int__(self):
...
+                from .symbol import Dummy
+                isign = 1 if i > 0 else -1
+                x = Dummy()
+                # in the following (self - i).evalf(2) will not always work while
+                # (self - r).evalf(2) and the use of subs does; if the test that
+                # was added when this comment was added passes, it might be safe
+                # to simply use sign to compute this rather than doing this by hand:
+                diff_sign = 1 if (self - x).evalf(2, subs={x: i}) > 0 else -1
+                if diff_sign != isign:
+                    i -= isign
```
Comment (smichr) [inline sympy/core/expr.py:331]:
```suggestion
        if not i:
            return i
        if int_valued(r):
            # non-integer self should pass one of these tests
            if (self > i) is S.true:
                return i
            if (self < i) is S.true:
                return i - 1
            # off by one
            return i - (1 if i > 0 else -1)
```

**Comment by smichr** [comment]
> Shouldn't the code in round be handling this? The checks here imply that it's round that isn't correct.

Consider the test case `eps = S(1)/10**120; n = cos(2)**2+sin(2)**2 - eps`. `n` is nearly 1 so it should round to 1.0 but this is not the right int value and neither `n > 1` nor `n < 1` evaluate: we are off by one.

File: sympy/core/expr.py [hunk 9]
```diff
@@ -325,18 +312,15 @@ def __int__(self):
...
-            if diff_sign != isign:
-                i -= isign
+            return i
+        if int_valued(r):
+            # non-integer self should pass one of these tests
+            if (self > i) is S.true:
+                return i
+            if (self < i) is S.true:
+                return i - 1
+            # off by one
```
Comment (smichr) [inline sympy/core/expr.py:322]:
```suggestion
            ok = self.equals(i)
            if ok is None:
                raise TypeError('cannot compute int value accurately')
            if ok:
                return i
            # off by one
```

**Comment by smichr** [comment]
@asmeurer , have your questions been resolved?

**Pull request closed without merging**